### PR TITLE
ipq40xx: add support for Netgear EX6100v2/EX6150v2

### PR DESCRIPTION
**Pull request by blocktrron**
This adds support for the EX6100v2 and EX6150v2 Range-Extenders by Netgear.

### CODE REVIEW

**Comment by ecsv** [comment]
Looks to me like there are a lot of things which have to be adjusted to use the same style as the other devices. You can have a look at:

* https://github.com/openwrt/openwrt/commit/d0e962140403eefbe70002ee4349ade1d8c119be#diff-56084c5e657badcdff6ce29181444b61
* https://github.com/openwrt/openwrt/pull/810
* https://github.com/openwrt/openwrt/pull/713#issuecomment-369511149

Other things:

* Is there a reason why you marked ess-switch as gpio-controller?
* Why do the dts files only contain the copyright "2015, The Linux Foundation. All rights reserved." and not yours?
* is the u-boot requesting the naming of the reserved-memory node or could you also split it into the "smem"/"tz" regions as suggested by the ap-dk01 patch 156f2df879f
("ipq806x: ipq4019: add ap-dk01.1-c1 board support")?

File: target/linux/ipq40xx/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+PHYNBR=${DEVPATH##*/phy}
+
+[ -n $PHYNBR ] || exit 0
+
+. /lib/functions.sh
+. /lib/functions/system.sh
+
+board=$(board_name)
+
+case "$board" in
```
Comment (mkresin) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac]: drop this file and patch the wireless macs via https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata.

Have a look at the meraki,mr33. It should work the same way.

**Comment by blocktrron** [comment]
Thanks for your feedback!

@ecsv 
I've updated the device-trees of both devices to align the style to your linked commits. I've also combined the board-files of both devices into one single binary.

@mkresin 
Your proposed method of patching the MAC-Address indeed works like a charm. Thank you for that!

File: target/linux/ipq40xx/base-files/etc/board.d/02_network [hunk 2]
```diff
@@ -35,6 +35,10 @@ glinet,gl-b1300)
 	ucidef_add_switch "switch0" \
 		"0u@eth0" "3:lan" "4:lan"
 	;;
+netgear,ex6100v2 |\
```
Comment (mkresin) [inline target/linux/ipq40xx/base-files/etc/board.d/02_network]: merge this block with the meraki,mr33.

**Review by mkresin — CHANGES_REQUESTED** [review]

File: target/linux/ipq40xx/base-files/etc/board.d/02_network (hunk 2 above)
Comment (blocktrron) [reply target/linux/ipq40xx/base-files/etc/board.d/02_network]: fixed

**Comment by mkresin** [comment]
@ecsv this one looks good to me. Do you see anything missing/wrong?

**Comment by ecsv** [comment]
@blocktrron The variant string seems to be incorrect. The ones sent to Kalle Valo/ath10k in http://ath10k.infradead.narkive.com/geQgjTPc/qca4019-hw1-0-bdfs-for-netgear-ex6100v2-and-ex6150v2 use following strings to identify the BDFs:

* bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=Netgear-EX6100v2
* bus=ahb,bmi-chip-id=0,bmi-board-id=17,variant=Netgear-EX6100v2
* bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=Netgear-EX6150v2
* bus=ahb,bmi-chip-id=0,bmi-board-id=17,variant=Netgear-EX6150v2

But you are now using the variant strings with upper case "NETGEAR" in this pull request.

Are you sure that the tz_apps regions doesn't require a reserved memory region (maybe you have some documentation about that)?

What is the reason why you set "status = "okay";" for wifi0/1 in both dtsi and dts?

You don't seem to set the top control status registers for some of the enabled peripherals. Following things might be relevant for you:

* qcom,ess-interface-select
* qcom,wifi_glb_cfg
* qcom,wifi_noc_memtype_m0_m2

The original DT from http://www.downloads.netgear.com/files/GDC/EX6150v2/EX6150v2_EX6100v2-V1.0.1.54.zip (there are 9 and I don't know which one is loaded for which device - but only fdt@7 uses a qcom,ess-interface-select TCSR_ESS_PSGMII_RGMII4 instead of TCSR_ESS_PSGMII - doesn't seem to be interesting because you said that config@4+fdt@4 is the one which u-boot selects):

    tcsr@1949000 {
    	compatible = "qcom,tcsr";
    	reg = <0x1949000 0x100>;
    	qcom,wifi_glb_cfg = <TCSR_WIFI_GLB_CFG>;
    };
    
    ess_tcsr@1953000 {
    	compatible = "qcom,tcsr";
    	reg = <0x1953000 0x1000>;
    	qcom,ess-interface-select = <TCSR_ESS_PSGMII>;
    };
    
    tcsr@1957000 {
    	compatible = "qcom,tcsr";
    	reg = <0x1957000 0x100>;
    	qcom,wifi_noc_memtype_m0_m2 = <TCSR_WIFI_NOC_MEMTYPE_M0_M2>;
    };

It might be confusing that the LEDs on the ex6100v2 are called ex6150v2:*. What do you think @mkresin ? This should break for ex6100v2 because you use following in 01_leds:

    +netgear,ex6100v2 |\
    +netgear,ex6150v2)
    +	ucidef_set_led_wlan "wlan2g" "WLAN2G" "${boardname}:green:router" "phy0tpt"
    +	ucidef_set_led_wlan "wlan5g" "WLAN5G" "${boardname}:green:client" "phy1tpt"
    +	;;

Are we sure that the devices don't have an (internal) usb port/NGFF/USB-only-mini-PCIe..? Because the original DT enabled  the USB stuff and configured the tcsr for it. It is not very likely (didn't find anything on the product pictures) but maybe you can check the board.

**Comment by mkresin** [comment]
Once more, thanks a lot for pointing out the remaining issues.

> It might be confusing that the LEDs on the ex6100v2 are called ex6150v2:*. What do you think @mkresin ? This should break for ex6100v2 because you use following in 01_leds:

IMHO the dtsi should be renamed from `ex61xxv2.dtsi` to `ex61x0v2.dtsi` and all occurrences of ex61xxv2 be replaced with ex61x0v2 to be as specific as possible.

Instead of creating two dts files, the LEDs can be named `ex61x0v2:<colour>:<function>`. It should be the best tradeoff.

**Comment by blocktrron** [comment]
Thanks for your quick reviews!

@ecsv 
I were not aware of the board files being already submitted. However, they don't seem to be included (yet). I've altered the naming of the variant in the board-binary and the dts.

Regarding tz_apps, for me it seems like it is not necessary to include it. The decompiled device tree contains
```
rsvd@87E00000 {
	reg = <0x87e00000 0x200000>;
	no-map;
};
```

The redundant activation of both wifis is an oversight by myself, it is fixed now.

The removal of the tcsr was an accident. I probably looked in the wrong dtsi file, i was convinced the tcsrs were initialized in `qcom-ipq4019.dtsi`, i've just double checked, indeed this is not the case. Also fixed now.

Good catch on the LED naming, altered to use `ex61x0v2` as proposed by @mkresin now.

The device does not make any usage of USB Ports at all. Maybe you are able to solder one to the board, but IMHO this is out of scope here.

**Comment by mkresin** [comment]
@blocktrron please squash your commits.

**Comment by blocktrron** [comment]
@mkresin done

**Comment by mkresin** [comment]
merged, thanks!

**Pull request closed without merging**